hxunpipe.1
.de d \" begin display
.sp
.in +4
.nf
..
.de e \" end display
.in -4
.fi
.sp
..
.TH "HXUNPIPE" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxunpipe \- convert output of hxpipe back to XML format
.SH SYNOPSIS
.B hxunpipe
.RB "[\| " \-b " \|]"
.RI "[\| " file-or-URL " \|]"
.SH DESCRIPTION
.B hxunpipe
takes the output of
.BR hxpipe (1)
(or of 
.BR onsgmls (1))
and turns it back into XML/SGML mark-up.
.SH OPTIONS
The following options are supported:
.TP 10
.B \-b
Normally,
.B hxunpipe
assumes the input was made by hxpipe, i.e., the
input may contain character entities but will never contain SGML/XML
delimiters (<>&"') that need to be escaped. When the input was made by
(o)nsgmls, however, the entities will have been expanded and the input
may contain SGML/XML delimiters. The option
.B \-b
causes
.B hxunpipe
to look for those delimiters and escape them: "<" as "&lt;", ">" as
"&gt;", "&" as "&amp;", """ as "&quot;" and "'" as "&apos;".
.SH OPERANDS
The following operand is supported:
.TP 10
.I file-or-URL
The name or URL of an HTML file. If absent, standard input is read
instead.
.SH "EXIT STATUS"
The following exit values are returned:
.TP 10
.B 0
Successful completion.
.TP
.B > 0
An error occurred in the input.
.SH ENVIRONMENT
To use a proxy to retrieve remote files, set the environment variables
.B http_proxy
and
.BR ftp_proxy "."
E.g.,
.B http_proxy="http://localhost:8080/"
.SH BUGS
.LP
Not all syntax errors in the input are recognized.
.LP
.B hxunpipe
can currently only retrieve remote files over HTTP. It doesn't handle
password-protected files, nor files whose content depends on HTTP
"cookies."
.SH "SEE ALSO"
.BR hxpipe (1),
.BR onsgmls (1).


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxunpipe.c
/*
 * unpipe - takes output of pipe and convert to HTML/XML form
 *
 * Copyright © 1994-2000 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/copyright-software
 *
 * Author: Bert Bos <bert@w3.org>
 * Created: 23 May 1999
 * Version: $Id: hxunpipe.c,v 1.11 2017/11/24 09:50:25 bbos Exp $
 */
#include "config.h"
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <stdbool.h>
#ifdef HAVE_UNISTD_H
#  include <unistd.h>
#endif
#include <assert.h>
#ifdef HAVE_STRING_H
#  include <string.h>
#elif HAVE_STRINGS_H
#  include <strings.h>
#endif
#include "export.h"
#include "types.e"
#include "heap.e"
#include "errexit.e"
#include "dict.e"
#include "openurl.e"

static int nrattrs = 0;
static char **attrs = NULL;
static bool escape = false;

/* put_text -- replace newlines and print text */
static void put_text(FILE *in)
{
  int c, c1, c2;

  while ((c = getc(in)) != EOF && c != '\n')
    if (c != '\\') {
      if (!escape) putchar(c);
      else if (c == '<') printf("&lt;");
      else if (c == '>') printf("&gt;");
      else if (c == '"') printf("&quot;");
      else if (c == '&') printf("&amp;");
      else if (c == '\'') printf("&apos;");
      else putchar(c);
    }
    else if ((c = getc(in)) == EOF) return; /* Error */
    else if (c == '\n') return;		    /* Error */
    else if (c == 'n') putchar('\n');
    else if (c == 'r') putchar('\r');
    else if (c == 't') putchar('\t');
    else if (c == '#') printf("&#");
    else if (c < '0' || '7' < c) putchar(c);
    else if ((c1 = getc(in)) < '0' || '7' < c1) printf("%c%c", c, c1);
    else if ((c2 = getc(in)) < '0' || '7' < c2) printf("%c%c%c", c, c1, c2);
    else printf("&#%d;", 64 * (c - '0') + 8 * (c1 - '0') + (c2 - '0'));
}

/* store_attr -- store attributes temporarily */
static void store_attr(FILE *in)
{
# define INC 1014
  int c, n = 0;

  renewarray(attrs, nrattrs + 1);
  attrs[nrattrs] = NULL;
  renewarray(attrs[nrattrs], INC);
  while ((c = getc(in)) != EOF && c != '\n') {
    attrs[nrattrs][n++] = c;
    if (n % INC == 0) {renewarray(attrs[nrattrs], INC * (n/INC + 1));}
  }
  attrs[nrattrs][n] = '\0';
  nrattrs++;
}

/* put_attr -- write out attributes */
static void put_attr(void)
{
  int i, j;

  for (j = 0; j < nrattrs; j++) {
    for (i = 0; attrs[j][i] && attrs[j][i] != ' '; i++);
    if (attrs[j][i] != ' ') errexit("Incorrect A (attribute) line\n");
    if (! eq(attrs[j] + i + 1, "IMPLIED")) {
      putchar(' ');
      for (i = 0; attrs[j][i] && attrs[j][i] != ' '; i++) putchar(attrs[j][i]);
      if (attrs[j][i] != ' ') errexit("Incorrect A (attribute) line\n");
      putchar('=');
      for (i++; attrs[j][i] && attrs[j][i] != ' '; i++) ; /* skip type */
      if (attrs[j][i] != ' ') errexit("Incorrect A (attribute) line\n");
      putchar('"');
      for (i++; attrs[j][i]; i++) {
	if (attrs[j][i] != '\\') putchar(attrs[j][i]);
	else if (attrs[j][i+1]) {
	  i++;
	  if (attrs[j][i] == 'n') putchar('\n');
	  else if (attrs[j][i] == 'r') putchar('\r');
	  else if (attrs[j][i] == 't') putchar('\t');
	  else if (attrs[j][i] == '#') printf("&#");
	  else if ('0' <= attrs[j][i] && attrs[j][i] <= '7' &&
		   '0' <= attrs[j][i+1] && attrs[j][i+1] <= '7' &&
		   '0' <= attrs[j][i+2] && attrs[j][i+2] <= '7') {
	    printf("&#%d;", 64 * (attrs[j][i] - '0') +
		   8 * (attrs[j][i+1] - '0') + (attrs[j][i+2] - '0'));
	    i += 2;
	  } else putchar(attrs[j][i]);
	}
      }
      putchar('"');
    }
    dispose(attrs[j]);
  }
  nrattrs = 0;
}

/* put_decl -- write a DOCTYPE declaration */
static void put_decl(FILE *in)
{
  int c;
  bool hasfpi = false;

  printf("<!DOCTYPE ");

  /* Write name of root element */
  while ((c = getc(in)) != EOF && c != '\n' && c != ' ') putchar(c);

  /* Write FPI if present */
  while (c == ' ') c = getc(in);
  if (c == '"') {
    if ((c = getc(in)) == EOF || c == '\n')
      errexit("Incorrect DOCTYPE declaration\n");
    if (c != '"') {
      hasfpi = true;
      printf(" PUBLIC \"%c", c);
      while ((c = getc(in)) != EOF && c != '\n' && c != '"') putchar(c);
      if (c != '"') errexit("Incorrect DOCTYPE declaration\n");
      putchar('"');
    }
    c = getc(in);
  }

  /* Write URL if present */
  while (c == ' ') c = getc(in);
  if (c != EOF && c != '\n') {
    if (hasfpi) printf(" \"%c", c); else printf(" SYSTEM \"%c", c);
    while ((c = getc(in)) != EOF && c != '\n') putchar(c);
    putchar('"');
  }

  putchar('>');
}

/* usage -- print usage message and exit */
static void usage(string prog)
{
  fprintf(stderr, "Version %s\nUsage: %s [file_or_url]\n", VERSION, prog);
  exit(1);
}

int main(int argc, char *argv[])
{
  int c, status = 200;
  FILE *in = NULL;
  bool empty = false;

  while ((c = getopt(argc, argv, "b")) != -1)
    switch (c) {
      case 'b': escape = true; break;
      default: usage(argv[0]);
    }
  if (optind == argc) in = stdin;
  else if (optind == argc - 1) in = fopenurl(argv[optind], "r", &status);
  else usage(argv[0]);

  if (in == NULL) { perror(argv[optind]); exit(2); }
  if (status != 200) errexit("%s : %s\n", argv[optind], http_strerror(status));

  while ((c = getc(in)) != EOF) {
    switch (c) {
    case '-': put_text(in); break;
    case '?': printf("<?"); put_text(in); printf(">"); break;
    case '_': case '*': printf("<!--"); put_text(in); printf("-->"); break;
    case 'L': break;
    case 'A': store_attr(in); break;
    case '(': putchar('<'); put_text(in); put_attr(); putchar('>'); break;
    case ')':
      if (!empty) {printf("</"); put_text(in); putchar('>');}
      else empty = false;
      break;
    case '|': putchar('<'); put_text(in); put_attr(); printf(" />"); break;
    case '!': put_decl(in); break;
    case 'e': empty = true; break; /* Generated by onsgmls */
    case 'i': case 'o': break;	   /* Generated by onsgmls */
    case 'C': break;
    }
  }
  if (! feof(in)) { perror(argv[0]); exit(1); }
  fclose(in);
  return 0;
}
